File: ABNLookupSampleCodeVBDotNet/ABRXMLSearchClient/XMLMessageParser.vb
Imports System.Xml
Imports System.IO

Public Class XMLMessageParser
   ' Log for the message parser
   Private _xmlDoc As XmlDocument
   Public Property XmlDoc() As XmlDocument
      Get
         If _xmlDoc Is Nothing Then
            _xmlDoc = New XmlDocument
         End If
         Return _xmlDoc
      End Get
      Set(ByVal value As XmlDocument)
         _xmlDoc = value
      End Set
   End Property
   Private _namespaceManager As XmlNamespaceManager
   Public Property NamespaceManager() As XmlNamespaceManager
      Get
         If _namespaceManager Is Nothing Then
            initialiseNamespace()
         End If
         Return _namespaceManager
      End Get
      Set(ByVal value As XmlNamespaceManager)
         _namespaceManager = value
      End Set
   End Property
   Public Sub New(ByVal xml As String)
      LoadXML(xml)
   End Sub
   Public Sub LoadXML(ByVal xml As String)
      Try
         XmlDoc.LoadXml(xml)
      Catch exp As Exception
         Throw
      End Try
   End Sub
   '---------------------------------------------------------------------------------------
   ' Get the name details from the dom
   '---------------------------------------------------------------------------------------
   Public Function GetNames() As DataTable
      Dim MatchingName As DataRow
      Dim MatchingNames As DataTable = CreateNewNamesDataTable()
      Dim Names As XmlNodeList
      Dim SoapException As XmlNode
      Try
         SoapException = XmlDoc.SelectSingleNode(XPath.LocationSOAP(XPath.Path.Exception), NamespaceManager)
         If SoapException Is Nothing Then
            Names = XmlDoc.SelectNodes(XPath.LocationSOAP(XPath.Path.commonRoot), NamespaceManager)
            For Each Name As XmlNode In Names
               MatchingName = MatchingNames.NewRow
               MatchingName("ABN") = getNodeValue(Name, XPath.LocationSOAP(XPath.Path.abnValue))
               MatchingName("Entity Name") = getNodeValue(Name, XPath.LocationSOAP(XPath.Path.name))
               MatchingName("State") = getNodeValue(Name, XPath.LocationSOAP(XPath.Path.postcode))
               MatchingName("Postcode") = getNodeValue(Name, XPath.LocationSOAP(XPath.Path.state))
               MatchingName("Score") = getNodeValue(Name, XPath.LocationSOAP(XPath.Path.score))
               MatchingNames.Rows.Add(MatchingName)
            Next
         Else
            Throw New System.Exception("Exception from Soap Search")
         End If
      Catch exp As Exception
         Throw exp
      End Try
      Return MatchingNames
   End Function
   '---------------------------------------------------------------------------------------
   ' Create  data table for the names collection 
   '---------------------------------------------------------------------------------------
   Private Function CreateNewNamesDataTable() As DataTable
      Dim MatchingNames As New DataTable
      MatchingNames = New DataTable
      MatchingNames.Columns.Add("ABN")
      MatchingNames.Columns.Add("Entity Name")
      MatchingNames.Columns.Add("State")
      MatchingNames.Columns.Add("Postcode")
      MatchingNames.Columns.Add("Score")
      Return MatchingNames
   End Function
   '---------------------------------------------------------------------------------------
   ' Add namespace to the namespace manager
   '---------------------------------------------------------------------------------------
   Private Sub initialiseNamespace()
      Const _soap As String = "SOAP"
      Const _namepaceAlias As String = "abn"
      Const _namepace As String = "http://abr.business.gov.au/ABRXMLSearch/"
      Me.NamespaceManager = New XmlNamespaceManager(Me.XmlDoc.NameTable)
      Me.NamespaceManager.AddNamespace(_namepaceAlias, _namepace)
   End Sub
   '---------------------------------------------------------------------------------------
   ' Add namespace to the namespace manager
   '---------------------------------------------------------------------------------------
   Private Function getNodeValue(ByVal node As XmlNode, ByVal xPath As String) As String
      Dim Element As XmlNode
      Element = node.SelectSingleNode(xPath, NamespaceManager)
      If Not Element Is Nothing Then
         Return Element.InnerText
      Else
         Return ""
      End If
   End Function

End Class

File: ABNLookupSampleCodeCSharp35/SampleAbnLookupClient/XmlMessageParser.cs
using System;
using System.Xml;
using System.IO;
using System.Data;
using System.Globalization;

namespace AbnLookup.SearchClientCSharpe {
   /// <summary>
   /// Process XML message using DOM.
   /// </summary>
   class XmlMessageParser {
      // ---------------------------------------------------------------------------------------
      // private constructor
      // ---------------------------------------------------------------------------------------
      private XmlMessageParser() { }
      // ---------------------------------------------------------------------------------------
      //  Get the name details from the dom
      // ---------------------------------------------------------------------------------------
      public static DataTable GetNames(string xml) {
         DataTable MatchingNames = CreateNewNamesDataTable();
         XmlDocument XmlDoc = LoadXml(xml);
         XmlNamespaceManager NamespaceManager = initialiseNamespace(XmlDoc);
         try {
            if (!IsException(XmlDoc, NamespaceManager)) {
               XmlNodeList Names = XmlDoc.SelectNodes(AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.CommonRoot), NamespaceManager);
               foreach (System.Xml.XmlNode Name in Names) {
                  DataRow MatchingName = MatchingNames.NewRow();
                  MatchingName["ABN"] = getNodeValue(Name, AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.AbnValue), NamespaceManager);
                  MatchingName["Entity Name"] = getNodeValue(Name, AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.Name), NamespaceManager);
                  MatchingName["State"] = getNodeValue(Name, AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.State), NamespaceManager);
                  MatchingName["Postcode"] = getNodeValue(Name, AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.Postcode), NamespaceManager);
                  MatchingName["Score"] = getNodeValue(Name, AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.Score), NamespaceManager);
                  MatchingNames.Rows.Add(MatchingName);
               }
            }
            else {
               throw new System.Exception("Exception from Soap Search");
            }
         }
         catch {
            throw;
         }
         return MatchingNames;
      }
      // ---------------------------------------------------------------------------------------
      // load xml into the dom
      // ---------------------------------------------------------------------------------------
      private static XmlDocument LoadXml(string xml) {
         XmlDocument XmlDoc = new XmlDocument();
         try {
            XmlDoc.LoadXml(xml);
         }
         catch {
            throw;
         }
         return XmlDoc;
      }
      // ---------------------------------------------------------------------------------------
      //  return true if there was an exception returned by the Soap call
      // ---------------------------------------------------------------------------------------
      private static bool IsException(XmlDocument xmlDoc, XmlNamespaceManager namespaceManager) {
         XmlNode SoapException;
         SoapException = xmlDoc.SelectSingleNode(AbnLookupXpath.LocationSoap(AbnLookupXpath.Path.Exception), namespaceManager);
         return !(SoapException == null);
      }
      // ---------------------------------------------------------------------------------------
      //  Create  data table for the names collection 
      // ---------------------------------------------------------------------------------------
      private static DataTable CreateNewNamesDataTable() {
         DataTable MatchingNames = new DataTable();
         MatchingNames.Locale = CultureInfo.CurrentCulture;
         MatchingNames.Columns.Add("ABN");
         MatchingNames.Columns.Add("Entity Name");
         MatchingNames.Columns.Add("State");
         MatchingNames.Columns.Add("Postcode");
         MatchingNames.Columns.Add("Score");
         return MatchingNames;
      }
      // ---------------------------------------------------------------------------------------
      //  Add namespace to the namespace manager
      // ---------------------------------------------------------------------------------------
      private static XmlNamespaceManager initialiseNamespace(XmlDocument xmlDoc) {
         const string NAMEPACE_ALIAS = "abn";
         const string NAMEPACE = "http://abr.business.gov.au/ABRXMLSearch/";
         XmlNamespaceManager NamespaceManager = new XmlNamespaceManager(xmlDoc.NameTable);
         NamespaceManager.AddNamespace(NAMEPACE_ALIAS, NAMEPACE);
         return NamespaceManager;
      }
      // ---------------------------------------------------------------------------------------
      //  Add namespace to the namespace manager
      // ---------------------------------------------------------------------------------------
      private static string getNodeValue(XmlNode node, string xPath, XmlNamespaceManager namespaceManager) {
         XmlNode Element;
         Element = node.SelectSingleNode(xPath, namespaceManager);
         if (!(Element == null)) {
            return Element.InnerText;
         }
         else {
            return "";
         }
      }
   }
}
